subprojects/language-java/src/main/java/org/gradle/api/tasks/javadoc/Javadoc.java

```diff
@@ -296,7 +298,7 @@ public void setClasspath(FileCollection classpath) {
      * @return The options. Never returns null.
      */
     @Nested
-    public MinimalJavadocOptions getOptions() {
+    public StandardJavadocDocletOptions getOptions() {
```

Comment:
There's a couple things here. We _could_ leave this as `MinimalJavadocOptions`, and still have all the annotations from `StandardJavadocDocletOptions` be picked up by declaring the private field so. I wonder why you the return type of the getter should not be changed to the new reality, though?

As to using runtime types: it would certainly be nice to use inspect the runtime types. We could for example handle collections and maps of inputs as well. And runtime type inspections is arguably more in-line with user expectations. It seems possible to implement this, but it's not trivial to do so.

We also do some validation in the plugin development plugin's `ValidateTaskProperties` task, which needs to base its decisions on the declared types alone. We could say that's acceptable, though. The risk would be that the user passes some value that declares more properties (or properties with different annotations), which could lead to confusing problems.